### PR TITLE
✨ [Mailer] Add Mailjet bridge

### PR DESCRIPTION
**Pull request by tcheymol**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | no
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/14095

### CODE REVIEW

File: src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+4.4.0
```
Comment (noniagriconomie) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md:4]: Should be 5.1 IMO

Comment (tcheymol) [reply src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md]: Okay, I did change that, thanks :)

Comment (tcheymol) [reply src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md]: I get a probem on the Travis I did not have before, is it something I can fix ?

**Comment by lwillems** [comment]
@tcheymol what about Api transport with mailjet ? i was working on it before i saw your PR so perhaps i can help if you have not done already ?

**Comment by tcheymol** [comment]
@lwillems Yes, I started working on it, but I needed a first working version quite quick, so, I did not work on the Api part at first

File: src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md (hunk 1 above)
Comment (noniagriconomie) [reply src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md]: i do not know, maybe make sure your branch is rebase on the master
and retry build?

**Comment by fabpot** [comment]
@tcheymol Any plan to add support for the Mailjet API? Perhaps @lwillems can help you finish the PR?

**Comment by tcheymol** [comment]
@fabpot Yes, that is my plan indeed :)

**Comment by fabpot** [comment]
@tcheymol Can you submit a PR on symfony/recipes (you can have a look at other notifier bridge recipes to get some inspiration)?

File: src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.1
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md:4]: Should be `5.2`.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Mailjet Bridge
+===============
+
+Provides Mailjet integration for Symfony Mailer.
+
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:5]: You can provide a DSN example.

**Review by YaFou — CHANGES_REQUESTED** [review]

**Comment by tcheymol** [comment]
@fabpot You are right, I will look at the recipes, I will work on this this week

File: src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md (hunk 2 above)
Comment (tcheymol) [reply src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md]: Done

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md (hunk 3 above)
Comment (tcheymol) [reply src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:5]: Done

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+Mailjet Bridge
+===============
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:2]: Extra `=`

File: src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.2
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/CHANGELOG.md:4]: Should be `5.2.0` to be consistent with other similar files.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+---------
+
+```dotenv
+# Api usage
+MAILER_DSN=mailjet+api://$PUBLIC_KEY:$PRIVATE_KEY@default
+# Smtp usage
+MAILER_DSN=mailjet+smtp://$PUBLIC_KEY:$PRIVATE_KEY@default
+```
+
+
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:18]: Extra blank line can be removed

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use function array_key_exists;
+use function array_map;
+use function count;
+use function implode;
+use function is_array;
+use function rewind;
+use function sprintf;
+use function stream_get_contents;
+use function stream_get_meta_data;
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:25]: All `use function` lines should be removed.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+class MailjetApiTransport extends AbstractApiTransport
+{
+    private const HOST = 'api.mailjet.com';
+    private const API_VERSION = '3.1';
+    private $privateKey;
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:31]: Can you add a blank line above this one?

**Comment by tcheymol** [comment]
@lwillems I have been working on API integration, if you want to help me with your experience, I would love it

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+                $inlines[] = $formattedAttachment;
+            } else {
+                $attachments[] = $formattedAttachment;
+            }
+        }
+
+        return [$attachments, $inlines, $html];
+    }
+
+    public function __toString(): string
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:139]: Can you move it just after `__construct()`?

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 10]
```diff
@@ -0,0 +1,148 @@
+<?php
+
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:2]: Missing license header

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+namespace Symfony\Component\Mailer\Bridge\Mailjet\Transport;
+
+
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:6]: extra blank line can be removed

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        $this->publicKey = $publicKey;
+        $this->privateKey = $privateKey;
+
+        parent::__construct($client, $dispatcher, $logger);
+    }
+
+    protected function doSendApi(SentMessage $sentMessage, Email $email, Envelope $envelope): ResponseInterface
+    {
+        $response = $this->client->request('POST', 'https://'.$this->getEndpoint().'/v'.self::API_VERSION.'/send', [
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:44]: You can use `sprintf` function for the URL.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/composer.json [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+      "Symfony\\Component\\Mailer\\Bridge\\Mailjet\\": ""
+    },
+    "exclude-from-classmap": [
+      "/Tests/"
+    ]
+  },
+  "minimum-stability": "dev",
+  "extra": {
+    "branch-alias": {
+      "dev-master": "5.1-dev"
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/composer.json:36]: 5.2-dev

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+            ],
+            'auth_basic' => $this->publicKey.':'.$this->privateKey,
+            'json' => $this->getPayload($email, $envelope),
+        ]);
+
+        $result = $response->toArray(false);
+
+        if (200 !== $response->getStatusCode()) {
+            if ('application/json' === $response->getHeaders(false)['content-type'][0]) {
+                throw new HttpTransportException('Unable to send an email: '.$result['message'].sprintf(' (code %d).', $response->getStatusCode()), $response);
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:56]: You should use `sprintf` function for the message.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+            }
+
+            throw new HttpTransportException('Unable to send an email: '.$response->getContent(false).sprintf(' (code %d).', $response->getStatusCode()), $response);
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:59]: Same here.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        // The response needs to contains a 'Messages' key that is an array
+        if (!array_key_exists('Messages', $result) || !is_array($result['Messages']) || 0 === count($result['Messages'])) {
+            throw new HttpTransportException('Unable to send an email: '.$response->getContent(false).' malformed api response', $response);
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:64]: Same here.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetTransportFactory.php [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        $scheme = $dsn->getScheme();
+        $user = $this->getUser($dsn);
+        $password = $this->getPassword($dsn);
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+
+        if ('maijlet+api' === $scheme) {
+            return (new MailjetApiTransport($user, $password, $this->client, $this->dispatcher, $this->logger))->setHost($host);
+        }
+
+        if ('mailjet+smtp' === $scheme || 'mailjet+smtps' === $scheme || 'mailjet' === $scheme) {
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetTransportFactory.php:32]: Use `in_array` function should be better.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/composer.json (hunk 13 above)
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/composer.json:36]: `5.2-dev`

File: src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+            $message['Bcc'] = implode(',', $this->stringifyAddresses($emails));
+        }
+        if ($email->getTextBody()) {
+            $message['TextPart'] = $email->getTextBody();
+        }
+        if ($html) {
+            $message['HTMLPart'] = $html;
+        }
+
+        $payload = [
```
Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:111]: Can be directly in the return statement.

Comment (YaFou) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php:111]: Can be directly in the return statement.

**Review by YaFou — CHANGES_REQUESTED** [review]

**Comment by tcheymol** [comment]
> @tcheymol Can you submit a PR on symfony/recipes (you can have a look at other notifier bridge recipes to get some inspiration)?

https://github.com/symfony/recipes/pull/809

By the way, how do you handle Packagist repositories creation ?

**Comment by fabpot** [comment]
> > @tcheymol Can you submit a PR on symfony/recipes (you can have a look at other notifier bridge recipes to get some inspiration)?
> 
> [symfony/recipes#809](https://github.com/symfony/recipes/pull/809)
> 
> By the way, how do you handle Packagist repositories creation ?

Not sure I understand the question, but the code is automatically extracted to its own repository which is then submitted to Packagist.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md [hunk 19]
```diff
@@ -0,0 +1,24 @@
+Mailjet Bridge
+==============
+
+Provides Mailjet integration for Symfony Mailer.
+
+
+
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:7]: Extra blank lines can be removed

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+Configuration examples :
+---------
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:9]: This line should be removed as it's not a title.

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+```dotenv
+# Api usage
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:12]: API

File: src/Symfony/Component/Mailer/Bridge/Mailjet/README.md [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+MAILER_DSN=mailjet+api://$PUBLIC_KEY:$PRIVATE_KEY@default
+# Smtp usage
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailjet/README.md:14]: SMTP

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
@tcheymol Can you fix the minor issues so that I can merge? Thank you.

**Comment by fabpot** [comment]
Thank you @tcheymol.

**Pull request merged**